File: tools/core070_bench_prerun.R
# M3 performance pre-run (D-139): small paired timing probe, frozen R side.
# NOT a campaign — 3 sizes x 2 families x 3 reps, single thread, to ground the
# full benchmark estimate. Comparison harness lives in tools/, outside tests.
#
# argv: <frozen-library> <output-dir>
args <- commandArgs(TRUE)
stopifnot(length(args) == 2L)
frozen_library <- normalizePath(args[[1]], mustWork = TRUE)
out_dir <- args[[2]]
dir.create(out_dir, recursive = TRUE, showWarnings = FALSE)
.libPaths(c(frozen_library, .libPaths()))
suppressMessages(library(gllvmTMB))

grid <- expand.grid(p = c(5L, 20L, 50L), n = c(200L, 500L), family = c("gaussian", "poisson"),
                    stringsAsFactors = FALSE)
grid <- grid[!(grid$p == 50L & grid$n == 200L), ]  # keep the probe small
reps <- 3L
rows <- list()
for (g in seq_len(nrow(grid))) {
  p <- grid$p[g]; n <- grid$n[g]; fam <- grid$family[g]
  set.seed(90000 + g)
  K <- 2L
  Lambda <- matrix(rnorm(p * K, 0, 0.5), p, K)
  eta <- Lambda %*% matrix(rnorm(K * n), K, n)
  Y <- if (fam == "gaussian") eta + 0.7 * matrix(rnorm(p * n), p, n) else
       matrix(rpois(p * n, exp(pmin(eta, 4))), p, n)
  tn <- paste0("t", seq_len(p))
  df <- data.frame(site = factor(rep(seq_len(n), each = p)),
                   trait = factor(rep(tn, n), levels = tn),
                   value = as.vector(Y))
  fam_obj <- if (fam == "gaussian") gaussian() else poisson()
  times <- numeric(reps); ll <- NA_real_
  for (r in seq_len(reps)) {
    t0 <- proc.time()[["elapsed"]]
    fit <- gllvmTMB(value ~ 0 + trait + latent(0 + trait | site, d = 2, unique = FALSE),
                    data = df, unit = "site", trait = "trait", family = fam_obj,
                    control = gllvmTMBcontrol(n_init = 1L, se = FALSE))
    times[r] <- proc.time()[["elapsed"]] - t0
    ll <- as.numeric(logLik(fit))
  }
  rows[[length(rows) + 1L]] <- data.frame(p = p, n = n, family = fam,
    median_s = median(times), min_s = min(times), loglik = ll)
  # persist the data for the Julia side to fit the identical matrix
  saveRDS(list(Y = Y, p = p, n = n, family = fam),
          file.path(out_dir, sprintf("bench-data-%02d.rds", g)))
  write.csv(Y, file.path(out_dir, sprintf("bench-Y-%02d.csv", g)), row.names = FALSE)
}
res <- do.call(rbind, rows)
res$engine <- "gllvmTMB-frozen"
write.csv(res, file.path(out_dir, "r-timings.csv"), row.names = FALSE)
manifest <- data.frame(case = seq_len(nrow(grid)), p = grid$p, n = grid$n,
                       family = grid$family)
write.csv(manifest, file.path(out_dir, "bench-manifest.csv"), row.names = FALSE)
print(res)

File: tools/core070_bench_prerun.jl
# M3 performance pre-run, Julia side: fit the identical matrices the R stage
# saved (bench-Y-*.csv + bench-manifest.csv) with the paired canonical model
# (trait-dummy X mean design, matching value ~ 0 + trait + latent(..., d=2,
# unique=FALSE)), timing 3 reps each. Same convention as the ns2 batch.
# argv: ARGS[1] = the pre-run output dir written by core070_bench_prerun.R
using GLLVM
using DelimitedFiles
using Statistics

out = ARGS[1]
man, _ = readdlm(joinpath(out, "bench-manifest.csv"), ',', header = true)
_loglik(fit) = hasproperty(fit, :logLik) ? fit.logLik : fit.loglik

open(joinpath(out, "julia-timings.csv"), "w") do io
    println(io, "case,p,n,family,median_s,min_s,loglik")
    for i in axes(man, 1)
        case = Int(man[i, 1]); p = Int(man[i, 2]); n = Int(man[i, 3])
        fam = String(strip(string(man[i, 4])))
        raw, _ = readdlm(joinpath(out, "bench-Y-" * lpad(case, 2, '0') * ".csv"),
                         ',', header = true)
        Y = Matrix{Float64}(raw)
        @assert size(Y) == (p, n)
        X = zeros(p, n, p)
        for j in 1:p
            X[j, :, j] .= 1
        end
        times = Float64[]; ll = NaN
        for r in 1:3
            t = @elapsed begin
                # Poisson's per-trait intercepts β are intrinsic to the native
                # model (no X kwarg) and correspond to R's `0 + trait` means;
                # the Gaussian path takes the trait-dummy X tensor (ns2 convention).
                fit = fam == "poisson" ?
                    GLLVM.fit_poisson_gllvm(Int.(round.(Y)); K = 2) :
                    GLLVM.fit_gaussian_gllvm(Y; K = 2, X = X)
                ll = _loglik(fit)
            end
            push!(times, t)
        end
        println(io, join([case, p, n, fam, median(times), minimum(times), ll], ","))
        println(stderr, "done case=$case fam=$fam p=$p n=$n median=$(round(median(times), digits = 3))s ll=$ll")
    end
end
println("BENCH_PRERUN_JULIA_DONE")
